<a href="https://colab.research.google.com/github/Lolla-data-analyst/flyrank-ml-internship/blob/main/research_question_lane1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Research Question — Lane 1: Ranking Signal Analysis

## Research Question
Which pages are gaining search visibility (rising impressions) without a
matching rise in clicks, signaling a missed opportunity to capture that
visibility?

## Unit of Analysis
One row = one page's search performance.

## Output
A ranked list of pages where impressions are rising but CTR/clicks
aren't keeping pace, sorted with the biggest "missed opportunity" pages
at the top — possibly supported by a chart.

## Action
For pages near the top of the list, a content/SEO team member reviews
the page and adds a clearer call-to-action and product link, to help
convert existing visibility into clicks/engagement.

## Cost of Being Wrong
Wasted team effort on pages that didn't need attention, and more
seriously real declining pages get overlooked while attention is
misdirected elsewhere.

In [2]:
import os, subprocess

REPO_URL = "https://github.com/Lolla-data-analyst/flyrank-ml-internship"
REPO_DIR = "flyrank-ml-internship"

if not os.path.isdir(REPO_DIR):
    subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
os.chdir(REPO_DIR)

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "CSV not found — check clone worked"
print("Data found, ready to go.")

Working dir: /content/flyrank-ml-internship
Data found, ready to go.


In [3]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# 1. How many pages have real visibility (decent impressions)?
visible = df[df["impressions_90d"] >= 500]
print(f"Pages with 500+ impressions in 90 days: {len(visible)} out of {len(df)} total pages")

# 2. What's the average CTR among these visible pages?
avg_ctr = visible["ctr"].mean()
print(f"Average CTR among visible pages: {avg_ctr:.3f}")

# 3. How many visible pages have CTR notably below that average (e.g., half the average)?
underperforming = visible[visible["ctr"] < (avg_ctr * 0.5)]
print(f"Visible pages with CTR less than half the average: {len(underperforming)} "
      f"({len(underperforming)/len(visible)*100:.1f}% of visible pages)")

Pages with 500+ impressions in 90 days: 16726 out of 30000 total pages
Average CTR among visible pages: 0.262
Visible pages with CTR less than half the average: 7084 (42.4% of visible pages)


## Why This Lane Is Worth Pursuing

Using the starter dataset, 16,726 of 30,000 pages (55.8%) have meaningful visibility (500+ impressions in 90 days).
Among these, the average CTR is 0.262, but 7,084 pages (42.4% of visible pages) fall below half that average — showing a large, real
population of pages that are gaining visibility without converting it into clicks.
This confirms the research question is worth pursuing at scale, not just as an edge case.